AutomatticTracks/src/test/java/com/automattic/android/tracks/CrashLogging/CrashLoggingTest.kt

```diff
@@ -0,0 +1,70 @@
+package com.automattic.android.tracks.crashlogging
+
+import com.automattic.android.tracks.TracksUser
+import com.automattic.android.tracks.fakes.FakeDataProvider
+import com.automattic.android.tracks.fakes.testUser
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.never
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.times
+import org.mockito.Mockito.verify
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+
+@RunWith(RobolectricTestRunner::class)
+class CrashLoggingTest {
```

Comment:
After reviewing this class and leaving some suggestions about Kotlin usage, which I think can still be useful as a discussion point, I am failing to understand the testing approach here. These tests read to me as if it's trying to test the implementation details, but it doesn't have a way to do that, so it's trying to fake it. Let me explain.

First of all, the tests are extremely fragile. For example, if I were to add lines with just `dataProvider.applicationContext` and `dataProvider.userContext` to the `CrashLogging.setNeedsDataRefresh` method, I'd get 3 test failures. Of course having those lines by themselves would be meaningless because this would be a read-only action.

The issue is the tests are utilizing the `CrashLoggingDataProvider` interface which by design is just used to provide data. Whether that data is read once or 10 times doesn't matter. You can maybe change those to be `at least once`, in which case they'll provide some value, but I am not sure if that'd be enough or whether they'd match what the test names indicate.

I think the tests would be more useful if we had an access to an interface which prompted to do things. Think of names with actions in them, such as `applyUserTracking`. In that case, we could maybe write a test that says `if this happens make sure applyUserTracking is called once`.

In its current state, I feel like the tests are not there to help the implementation, so I think it's worth considering removing the tests entirely or re-working them to provide more value. I hope that makes sense!

Reply:
> For example, if I were to add lines with just dataProvider.applicationContext and dataProvider.userContext to the CrashLogging.setNeedsDataRefresh method, I'd get 3 test failures. 

It's good that you'd get test failures because those fields are not meant to be assigned anywhere in `CrashLogging#setNeedsDataRefresh`. I know it's not ideal (are requested but not assigned so eventually behavior won't change) but it at least indicates that something is changed in a way that was initially implemented.

> I think the tests would be more useful if we had access to an interface that prompted to do things

I get your points. In my opinion test in the current state does provide some value. Thinking of a scenario when some changes will be performed in `CrashLogging`, this test will assert that after changes, the fields that were previously requested, are still requested. Requesting a field is likely to equal setting this in Sentry. I know - not ideal.

I'm okay though to abstract `Sentry` static invocations to some class, this way we'll have a possibility to have more robust test but again, I don't think this test in its current form is problematic or doesn't bring a value.

Reply:
I see your point that the tests bring some value. At the same time I think the value is very low and I feel like these tests won't help us with refactoring and extending this class. As Oguz mentioned even a small unrelated change to the class breaks 3 tests. 

We are checking if certain fields were read but nothing else - eg. if I remove "Sentry.setUser(user)" the tests will still pass even though the code doesn't work.

IMO unit tests should be trying to verify that certain actions result in the expected result. I'm worried the current approach might make extending this class more difficult except of facilitating it.  I think the main issue is that the class is not testable at all - tbh the Sentry class is here to blame, since it makes testing way more difficult than it should. All in all, I'd consider removing these tests or trying to make the class testable (wrapping Sentry as you suggested) and writing a new unit test suite.

 > but it at least indicates that something is changed in a way that was initially implemented.

This might end up leading to a completely different discussion, so please feel free to ignore this part completely. But IMO this is something unit tests should never try to do. As mentioned above, IMO unit test should verify that certain action results in the expected result, but should not try to verify anything else. In might experience doing so can do more harm than good.

Reply:
I've majorly refactored this test to make it more robust and to check more things. I like it much better now, thank you for your feedback! You can check it here: https://github.com/Automattic/Automattic-Tracks-Android/pull/51/commits/713aeb434628356839e201ef3ef19ad828e7a987 but later I added so improvements so you may prefer to check the whole branch changes.

> IMO unit test should verify that certain action results in the expected result, but should not try to verify the internal implementation

I think we are both on the same page with that. I see I didn't present my thought as I wanted to: by saying "something is changed" I didn't mean an internal implementation but rather "something changed in behavior". Again - this wording was poorly on my side so sorry for the confusion.